src/vivarium_nih_moud/data/dismod_at.py

```diff
@@ -0,0 +1,342 @@
+    else:
+        assert 0, f'Method "{method}" unrecognized, should be "constant" or "linear"'
+    return f
+
+
+def data_model(name, f, df_data):  # FIXME: expose beta so it can be common across locations
+    if len(df_data) == 0:
+        return
+
+    ages = jnp.array(0.5 * (df_data.age_start + df_data.age_end))
+    years = jnp.array(0.5 * (df_data.year_start + df_data.year_end))
+    rate_obs_loc = jnp.array(df_data["mean"])
+    rate_obs_scale = jnp.array(df_data["standard_error"])
+
+    # refactor the following to use jax.vmap and to include population weights
+    rate_pred = jnp.zeros(len(df_data))
+    n_points = 5
+    for alpha in np.linspace(0, 1, n_points):
+        ages = jnp.array(alpha * df_data.age_start + (1 - alpha) * df_data.age_end)
+        rate_pred += f(ages, years)
+    rate_pred /= n_points
+
+    if len(df_data.filter(like="x_").columns) != 0:
+        # include fixed effects
+        X = jnp.array(df_data.filter(like="x_").fillna(0))
+        beta = numpyro.sample(
+            f"{name}_beta",
+            dist.Normal(loc=jnp.zeros(len(X[0])), scale=1.0),
+        )
+        rate_pred = jnp.exp(jnp.dot(X, beta)) * rate_pred
+
+    rate_obs = numpyro.sample(
+        f"{name}_obs", dist.Normal(loc=rate_pred, scale=rate_obs_scale), obs=rate_obs_loc
+    )
+    return rate_obs
+
+
+def at_param_w_data(param, ages, years, knot_val, df_data, method="constant"):
+    rate_function = at_param(param, ages, years, knot_val, method)
+    rate_data_obs = data_model(param, rate_function, df_data)
+    return rate_function
+
+
+def group_name(sex, location):
+    return f"{sex}_{location}".replace(" ", "_").lower()
+
+
+def single_location_model(
+    group,
+    sex,
+    location,
+    ages,
+    years,
+    knot_val_dict,
+    df_data,
+    include_consistency_constraints=True,
+):
+    group = group_name(sex, location)
+    i = at_param_w_data(
+        f"i_{group}", ages, years, knot_val_dict["i"], df_data[df_data.measure == "i"]
+    )
+    r = at_param_w_data(
+        f"r_{group}", ages, years, knot_val_dict["r"], df_data[df_data.measure == "r"]
+    )
+    f = at_param_w_data(
+        f"f_{group}", ages, years, knot_val_dict["f"], df_data[df_data.measure == "f"]
+    )
+    m = at_param_w_data(
+        f"m_{group}", ages, years, knot_val_dict["m"], df_data[df_data.measure == "m"]
+    )
+
+    p = at_param_w_data(
+        f"p_{group}",
+        ages,
+        years,
+        knot_val_dict["p"],
+        df_data[df_data.measure == "p"],
+        method="constant",
+    )
+
+    if include_consistency_constraints:
+        ode_model(group, p, i, r, f, m, sigma=0.01, ages=ages, years=years)
+    return dict(p=p, i=i, f=f, m=m, r=r)
+
+
+def ode_model(group, p, i, r, f, m, sigma, ages, years):
+    def dismod_f(t, y, args):
+        S, C = y
+        i, r, f, m = args
+        return (-m * S - i * S + r * C, -m * C + i * S - r * C - f * C)
```

Comment:
I object to the approach isort/black has taken to formatting this

Reply:
This is probably controversial because I know people are used to writing differential equations in a certain way, but if this is going to be code maybe you should make each item in this tuple a variable, and give everything here meaningful names instead of single letters?